File: combinatorics/resource/LCHP/problem_LTDT_code/Bai_1/bai1.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<vector<int>> partitions;

void printFerrers(const vector<int>& partition) {
    cout << "Phân hoạch: ";
    for (int i = 0; i < partition.size(); i++) {
        cout << partition[i];
        if (i < partition.size() - 1) cout << " + ";
    }
    cout << "\n";
    
    // In biểu đồ Ferrers gốc
    cout << "Biểu đồ Ferrers:\n";
    for (int part : partition) {
        for (int j = 0; j < part; j++) {
            cout << "*";
        }
        cout << "\n";
    }
    
    // In biểu đồ Ferrers chuyển vị
    cout << "Biểu đồ Ferrers chuyển vị:\n";
    if (!partition.empty()) {
        int maxCol = partition[0];  // Phần tử đầu tiên (lớn nhất)
        for (int col = 0; col < maxCol; col++) {
            for (int row = 0; row < partition.size(); row++) {
                if (col < partition[row]) {
                    cout << "*";
                }
            }
            cout << "\n";
        }
    }
    cout << "------------------------\n";
}

void generatePartitions(int n, int k, int currentSum, int lastValue, vector<int>& currentPartition) {
    // Nếu đã có đủ k phần
    if (currentPartition.size() == k) {
        if (currentSum == n) {
            partitions.push_back(currentPartition);
            printFerrers(currentPartition);
        }
        return;
    }
    
    // Nếu còn thiếu phần nhưng tổng đã bằng n
    if (currentSum == n) return;
    
    // Thử các giá trị từ 1 đến min(lastValue, n - currentSum)
    int remaining = n - currentSum;
    int remainingParts = k - currentPartition.size();
    
    // Đảm bảo có thể phân chia được
    for (int i = 1; i <= min(lastValue, remaining - remainingParts + 1); i++) {
        currentPartition.push_back(i);
        generatePartitions(n, k, currentSum + i, i, currentPartition);
        currentPartition.pop_back();
    }
}

int main() {
    int n, k;
    cout << "Nhập n và k: ";
    cin >> n >> k;
    
    // Kiểm tra điều kiện hợp lệ
    if (k > n || k <= 0 || n <= 0) {
        cout << "Không có phân hoạch nào thỏa mãn!\n";
        return 0;
    }
    
    vector<int> currentPartition;
    cout << "Các phân hoạch của " << n << " với đúng " << k << " phần:\n";
    cout << "========================\n";
    
    generatePartitions(n, k, 0, n, currentPartition);
    
    cout << "Tổng cộng có " << partitions.size() << " phân hoạch.\n";
    
    return 0;
}

File: combinatorics/resource/LCHP/problem_LTDT_code/Bai_1/bai1.py
def print_ferrers(partition):
    """In biểu đồ Ferrers và biểu đồ chuyển vị"""
    print("Phân hoạch:", " + ".join(map(str, partition)))
    
    # In biểu đồ Ferrers gốc
    print("Biểu đồ Ferrers:")
    for part in partition:
        print("*" * part)
    
    # In biểu đồ Ferrers chuyển vị
    print("Biểu đồ Ferrers chuyển vị:")
    if partition:
        max_col = partition[0]  # Phần tử đầu tiên (lớn nhất)
        for col in range(max_col):
            line = ""
            for row in range(len(partition)):
                if col < partition[row]:
                    line += "*"
            print(line)
    print("-" * 30)

def generate_partitions(n, k, current_sum=0, last_value=None, current_partition=None):
    """Sinh tất cả phân hoạch của n với đúng k phần"""
    if current_partition is None:
        current_partition = []
    if last_value is None:
        last_value = n
    
    # Nếu đã có đủ k phần
    if len(current_partition) == k:
        if current_sum == n:
            print_ferrers(current_partition[:])
            return [current_partition[:]]
        return []
    
    # Nếu tổng đã bằng n nhưng chưa đủ phần
    if current_sum == n:
        return []
    
    result = []
    remaining = n - current_sum
    remaining_parts = k - len(current_partition)
    
    # Thử các giá trị từ 1 đến min(last_value, remaining - remaining_parts + 1)
    for i in range(1, min(last_value, remaining - remaining_parts + 2)):
        if remaining >= i:
            current_partition.append(i)
            result.extend(generate_partitions(n, k, current_sum + i, i, current_partition))
            current_partition.pop()
    
    return result

def count_partitions_k_parts(n, k):
    """Đếm số phân hoạch của n với đúng k phần (không in ra)"""
    def dp_count(n, k, max_val):
        if k == 0:
            return 1 if n == 0 else 0
        if n <= 0 or max_val <= 0:
            return 0
        
        count = 0
        for i in range(1, min(max_val, n) + 1):
            count += dp_count(n - i, k - 1, i)
        return count
    
    return dp_count(n, k, n)

def main():
    try:
        n = int(input("Nhập n: "))
        k = int(input("Nhập k: "))
        
        # Kiểm tra điều kiện hợp lệ
        if k > n or k <= 0 or n <= 0:
            print("Không có phân hoạch nào thỏa mãn!")
            return
        
        print(f"Các phân hoạch của {n} với đúng {k} phần:")
        print("=" * 40)
        
        partitions = generate_partitions(n, k)
        
        print(f"Tổng cộng có {len(partitions)} phân hoạch.")
        
        # So sánh với công thức tính nhanh
        quick_count = count_partitions_k_parts(n, k)
        print(f"Kiểm tra bằng công thức: p_{k}({n}) = {quick_count}")
        
    except ValueError:
        print("Vui lòng nhập số nguyên hợp lệ!")

if __name__ == "__main__":
    main()